jax/experimental/pjit.py

```diff
@@ -0,0 +1,212 @@
+  return _pjit_callable(
+      fun, in_axis_resources, out_axis_resources_thunk, resource_env,
+      donated_invars, name, *in_avals)(*args)
+
+def _pjit_translation_rule(c, axis_env, in_nodes, name_stack, backend, name,
+                           call_jaxpr, in_axis_resources, out_axis_resources_thunk,
+                           resource_env, donated_invars):
+  mesh = resource_env.physical_mesh
+  subc = xc.XlaBuilder(f"pjit_{name}")
+
+  args = []
+  for i, (n, axis_resources) in enumerate(safe_zip(in_nodes, in_axis_resources)):
+    # N.B. inlined calls shouldn't have shardings set directly on the inputs or
+    # outputs (set_sharding_proto adds an identity operation).
+    arg = xb.parameter(subc, i, c.GetShape(n))
+    args.append(xb.set_sharding_proto(subc, arg,
+                                      get_sharding_proto(c, n, axis_resources, mesh)))
+
+  out_nodes = xla.jaxpr_subcomp(
+      subc, call_jaxpr, backend, axis_env, (),
+      extend_name_stack(name_stack, wrap_name(name, "pjit")), *args)
+  out_axis_resources = out_axis_resources_thunk()
+  out_nodes = [xb.set_sharding_proto(subc, out,
+                                     get_sharding_proto(c, n, axis_resources, mesh))
+               for out, axis_resources in safe_zip(out_nodes, out_axis_resources)]
+
+  subc = subc.build(xops.Tuple(subc, out_nodes))
+  return xops.Call(c, subc, list(in_nodes))
+
+pjit_call_p = core.CallPrimitive("pjit_call")
+pjit_call_p.def_impl(_pjit_call_impl)
+xla.call_translations[pjit_call_p] = _pjit_translation_rule
+
+# None indicates unpartitioned dimension
+ArrayAxisPartitioning = Union[pxla.MeshAxisName, Tuple[pxla.MeshAxisName, ...], None]
+# None indicates fully replicated array value
+ArrayPartitioning = Optional[Tuple[ArrayAxisPartitioning, ...]]
+
+@lu.cache
+def _pjit_callable(
+    fun: lu.WrappedFun,
+    in_axis_resources: Tuple[ArrayPartitioning, ...],
+    out_axis_resources_thunk: Callable[[], Tuple[ArrayPartitioning, ...]],
+    resource_env,
+    donated_invars,
+    name: str,
+    *in_avals):
+
+  in_axes = [get_array_mapping(axes) for axes in in_axis_resources]
+  out_axes_thunk = lambda: [get_array_mapping(axes)
+                            for axes in out_axis_resources_thunk()]
+  # TODO(skye): allow for using a submesh of physical_mesh
+  return pxla.mesh_callable(fun, name, None, resource_env.physical_mesh,
+                            in_axes, out_axes_thunk, donated_invars,
+                            True, *in_avals, tile_by_mesh_axes=False)
+
+
+def with_sharding_constraint(x, axis_resources):
+  resource_env = maps.thread_resources.env
+  return sharding_constraint_p.bind(x, axis_resources=axis_resources,
+                                    resource_env=resource_env)
+
+def _sharding_constraint_impl(x, axis_resources, resource_env):
+  # TODO(skye): can we also prevent this from being called in other
+  # non-pjit contexts? (e.g. pmap, control flow)
```

Comment:
Is pjit going to work eventually inside pmap, or xmap?

Reply:
Good question! I thought so, but after our discussion with James and Anselm, maybe pjit-of-vmap is enough. We should think more about this.